lib/baby_squeel/join_dependency.rb

```diff
@@ -35,43 +35,36 @@ def initialize(relation)
       # a list (in order of chaining) of associations and finding
       # the respective JoinAssociation at each level.
       def find_alias(associations)
-        table = find_join_association(associations).table
-        reconstruct_with_type_caster(table, associations)
+        # If we tell join_dependency to construct its tables, Active Record
+        # handles building the correct aliases and attaching them to its
+        # JoinDepenencies.
+        join_dependency.send(:construct_tables!, join_dependency.send(:join_root))
+        join_association = find_join_association(associations)
+
+        join_association.table
       end
 
       private
 
       def find_join_association(associations)
-        associations.inject(join_dependency.send(:join_root)) do |parent, assoc|
-          parent.children.find do |join_association|
-            reflections_equal?(
-              assoc._reflection,
-              join_association.reflection
-            )
-          end
-        end
-      end
-
-      # Compare two reflections and see if they're the same.
-      def reflections_equal?(a, b)
-        comparable_reflection(a) == comparable_reflection(b)
-      end
-
-      # Get the parent of the reflection if it has one.
-      # In AR4, #parent_reflection returns [name, reflection]
-      # In AR5, #parent_reflection returns just a reflection
-      def comparable_reflection(reflection)
-        [*reflection.parent_reflection].last || reflection
+        join_root = join_dependency.send(:join_root)
+        steps = associations.map {|a| a._reflection.name}
+        association_from_steps(join_root, steps)
       end
 
-      # Active Record 5's AliasTracker initializes Arel tables
-      # with the type_caster belonging to the wrong model.
+      # Search depth-first through the descendants of +current+ for successive
+      # associations in +steps+.
       #
-      # See: https://github.com/rails/rails/pull/27994
-      def reconstruct_with_type_caster(table, associations)
-        return table if ::ActiveRecord::VERSION::MAJOR < 5
-        type_caster = associations.last._scope.type_caster
-        ::Arel::Table.new(table.name, type_caster: type_caster)
+      # Each generation of descent consumes one element of +steps+.
+      def association_from_steps(current, steps)
+        return current if steps.empty?
+        matching_assoc = nil
+        current.children.each do |child|
+          next unless child.reflection.name == steps.first
+          matching_assoc = association_from_steps(child, steps[1..-1])
+          break if matching_assoc
+        end
+        matching_assoc
       end
```

Comment:
I haven't benchmarked this, but writing your logic like this could reduce iterations:

```ruby
def find_join_association(associations)
  current = join_dependency.send(:join_root)

  associations.each do |association|
    name = association._reflection.name
    current = current.children.find { |c| c.reflection.name == name }
  end

  current
end
```

However, I don't think comparing reflections by name alone is reliable enough. I haven't thought about this problem in quite a while, so my memory is foggy. Couldn't you have two models that each had an association named `:user`? 

Might also want to also check `c.reflection.klass == association._reflection.klass`. That won't work for polymorphic associations though. For that, you might find this comment useful: https://github.com/rzane/baby_squeel/issues/108#issuecomment-684962140

Reply:
I'll look over the "recursion -> iteration" optimization presently. Thanks for pointing that out.  I think it's a bit different than what you've given here, because there's a short-circuit exit from the recursion I don't see in your iterative implementation.  You can expect a commit from me soon on this.

As to name conflicts, this works through the association chain the same way the `BabySqueel::Association` was built.  As the code recurses through the `steps` (which I agree might go faster with iteration, instead), it finds associations by name effective _for that current model class_.  So, using the test model classes, it won't confuse `Comment#author` in `Post.where.has { comments.author.name == 'John' }` with `Post#author` because this code resolves `:comments` first, then looks in the `#children` for `:author`.  I thought that this would be a more efficient test than the full comparison of reflections and, since it took all the association steps into account, can't really be ambiguous -- but I'd welcome a unit test that proves me wrong, and I'll fix that 😉 .

Reply:
Indeed, there is a problem.  Here's my test code:
```ruby
      joined_relation = Post.joining { comments }
      relation = joined_relation.where.has { 
        c1 = author.name == 'Alice'
        c2 = comments.author.name == 'Bob'
        c1 & c2
      }
```
This is really spread out, so I can easily put in breakpoints.  My first guess is that the alias tracking doesn't carry over from `author.name` to `comment.author.name`.

Reply:
Huh, it'll be a miracle if I can make this work, since Active Record's `#where` doesn't get it right, either:
```ruby
Post.joins(:author => {}, :comments => {:author => {}}).where(:author => {:name => 'Alice'}, :comments => {:author => {:name => 'Bob'}}).to_sql
# => "SELECT \"posts\".* FROM \"posts\" INNER JOIN \"authors\" ON \"authors\".\"id\" = \"posts\".\"author_id\" INNER JOIN \"comments\" ON \"comments\".\"post_id\" = \"posts\".\"id\" INNER JOIN \"authors\" \"authors_comments\" ON \"authors_comments\".\"id\" = \"comments\".\"author_id\" WHERE \"author\".\"name\" = 'Alice' AND \"author\".\"name\" = 'Bob'"
```

(or more readable, without the Ruby syntax):
```sql
SELECT "posts".* FROM "posts"
INNER JOIN "authors" ON "authors"."id" = "posts"."author_id"
INNER JOIN "comments" ON "comments"."post_id" = "posts"."id"
INNER JOIN "authors" "authors_comments" ON "authors_comments"."id" = "comments"."author_id"
WHERE "author"."name" = 'Alice' AND "author"."name" = 'Bob'
```

☹️

Reply:
My commit below actually seems to fix the issue (and makes a great "selling point" for `baby_squeel`, since Active Record messes up the multiple aliases as I pointed out above).

Reply:
Actually, I think the solution to the issue I raised can be quite simple. An association can be uniquely identified by the model it is defined on + the name of the association.

If you look at the code before your changes, you'll see that's what it was doing.

Reply:
I think that is true only if there is no "loops" in the table relationships.  Take `Post.joining {[author, parent.author]}` as a counterexample: the `posts` and `authors` tables will be referenced twice, and the correct alias will need to be used in any further modifications to the `Relation` (`selecting`, `grouping`, etc.).